### PR TITLE
Fixes some problems with findPotentialStereo()

### PR DESCRIPTION
**Pull request by greglandrum**
My original version of `findPotentialStereo()` suffered from extreme under-testing and was pretty buggy.
Sorry to have created that mess. :-(

This fixes a bunch of problems/edge cases and adds a collection of new unit tests. 
In terms of regression testing: I've run this across 10K random ChEMBL molecules that have at least two chiral centers and compared the number of chiral centers and potential chiral centers found using `assignStereochemistry()` with the results from this new code. All the differences made sense and were expected.

The current implementation of the new tests is not super efficient. I have some ideas on how to speed it up, but unless there are strong objections I would prefer to get the fix in without delaying the release further and then start thinking about optimization. 

Out of an abundance of caution, the code is also now marked as experimental.

Note that this also fixes an edge-case bug in `rankMolAtoms()` which would only arise when the caller provides bond symbols (which is rarely done) and then only for some atom/bond orderings. 

Thanks to @ptosco for reporting the original problem that sent me down this rabbit hole.

### CODE REVIEW

**Review by ricrogz — COMMENTED** [review]
The code makes sense, and I can understand why these changes are required, but I'm kind of missing the global overview, and I can't for sure tell if there's anything else we are missing.

This looks good, except for my comments about that charges shouldn't create stereogenic centers, but I might be wrong on that.

File: Code/GraphMol/new_canon.cpp [hunk 1]
```diff
@@ -17,6 +17,7 @@
 #include <cstring>
 #include <iostream>
 #include <cassert>
+// #define VERBOSE_CANON 1
```
[the remark this answers is not in the document]
Comment (greglandrum) [reply Code/GraphMol/new_canon.cpp:20]: I leave that in there to make the next debugging session easier. :-)

File: Code/GraphMol/new_canon.cpp [hunk 2]
```diff
@@ -557,6 +558,10 @@ void updateAtomNeighborIndex(canon_atom *atoms, std::vector<bondholder> &nbrs) {
     nbr.nbrSymClass = newSymClass;
   }
   std::sort(nbrs.begin(), nbrs.end(), bondholder::greater);
+  // for (auto nbr : nbrs) {
+  //   std::cerr << "  " << nbr.nbrIdx << " " << nbr.nbrSymClass;
+  // }
+  // std::cerr << std::endl;
```
[the remark this answers is not in the document]
Comment (greglandrum) [reply Code/GraphMol/new_canon.cpp:564]: yep

File: Code/GraphMol/catch_chirality.cpp [hunk 3]
```diff
@@ -847,4 +899,320 @@ TEST_CASE("cleanup after removing possible centers", "[chirality]") {
...
+  SECTION("just a bug") {
+    // example that came up during testing
+
+    auto mol = "CC1=CN(C2OC(CNC(=O)C3c4ccccc4Sc4ccccc43)CC2)C(=O)NC1=O"_smiles;
+    REQUIRE(mol);
+    bool cleanIt = true;
+    bool flagPossible = true;
+    auto stereoInfo =
+        Chirality::findPotentialStereo(*mol, cleanIt, flagPossible);
+    CHECK(stereoInfo.size() == 2);
```
[the remark this answers is not in the document]
Comment (greglandrum) [reply Code/GraphMol/catch_chirality.cpp:1216]: Crap, yeah, that's fixed. I need to remove that comment.

File: Code/GraphMol/FindStereo.cpp [hunk 4]
```diff
@@ -289,11 +299,21 @@ std::string getBondSymbol(const Bond *bond) {
   return res;
 }
 
+namespace {
+std::string getAtomCompareSymbol(const Atom &atom) {
+  auto fmt = boost::format("%d%sH%d%d") % atom.getIsotope() % atom.getSymbol() %
+             atom.getTotalNumHs() % atom.getFormalCharge();
+  return fmt.str();
```
[the remark this answers is not in the document]
Comment (greglandrum) [reply Code/GraphMol/FindStereo.cpp:306]: Interesting. So `O[C@H][O-]F` wouldn't be chiral?

[the remark this answers is not in the document]
Comment (greglandrum) [reply Code/GraphMol/FindStereo.cpp:306]: I guess that's a bit hyper-artificial. Using the number of Hs is probably enough.

File: Code/GraphMol/catch_chirality.cpp [hunk 5]
```diff
@@ -299,6 +319,12 @@ TEST_CASE("isAtomPotentialTetrahedralCenter", "[unittest]") {
       CHECK(!Chirality::detail::isAtomPotentialTetrahedralCenter(
           mol->getAtomWithIdx(6)));
     }
+    {
+      auto mol = "O[P]([O-])(=O)OC"_smiles;
+      REQUIRE(mol);
+      CHECK(Chirality::detail::isAtomPotentialTetrahedralCenter(
+          mol->getAtomWithIdx(1)));
+    }
```
[the remark this answers is not in the document]
Comment (greglandrum) [reply Code/GraphMol/catch_chirality.cpp:327]: Hmm, ok. Maybe not for that one, but `O[C@H]([O-])F` does end up with a CIP code assigned

[the remark this answers is not in the document]
Comment (ricrogz) [reply Code/GraphMol/catch_chirality.cpp:327]: Hum, yes, the new labeler does label it... but the original "centres" program doesn't: it doesn't flag the C as stereogenic.

**Comment by bp-kelley** [comment]
Isn't the number of Hs kind of implied by charge?

**Comment by greglandrum** [comment]
> Isn't the number of Hs kind of implied by charge?

only if you ignore radicals

File: Code/GraphMol/catch_chirality.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (greglandrum) [reply Code/GraphMol/catch_chirality.cpp:327]: Ok, InChI agrees with you that neither the charge here nor that in N[C@]([NH3+])(F))Cl creates a chiral center. I will remove that.

To this specific comment though: the atom is still a _potential_ stereogenic center (which is what `isAtomPotentialTetrahedralCenter()` is testing) since it has degree 4. Once I fix the problem with charges it will no longer be identified as a result by `findPotentialStereo()` though.

**Comment by greglandrum** [comment]
Ok, suggested changes made.

**Review by ricrogz — APPROVED** [review]
Looks good now :)
